### PR TITLE
feat: app service plan autoscale based on memory percentage

### PR DESCRIPTION
**Pull request by MacMur85**
## Description

A custom autoscale policy for the app service plan using user-defined rules based on app metrics, so that the application can automatically scale based on its memory usage.

## Context

Instances Limits:

- Minimum: 1
- Maximum: 5
- Default: 1

Rule Scale Out:
- Metric Name: Memory Percentage
- Scale Trigger: > 70%
- Duration: 10 minutes
- Time Gain/Aggregation: Average
- Action: Increase the count by 1
- Cool Down: 5 minutes

Rule Scale In:
- Metric Name: Memory Percentage
- Scale Trigger: < 25%
- Duration: 10 minutes
- Time Gain/Aggregation: Average
- Action: Decrease the count by 1
- Cool Down: 5 minutes

## Type of changes

- [ ] Refactoring (non-breaking change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [X] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [X] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [X] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

**Comment by MacMur85** [comment]
Features deployed in the test environment
image
image

File: infrastructure/modules/app-service-plan/autoscale.tf [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  name                = "${var.names.app-service}-autoscale"
+  resource_group_name = var.resource_group_name
+  location            = var.location
+  target_resource_id  = azurerm_service_plan.appserviceplan.id
+
+  profile {
+    name = "${var.names.app-service}-MemoryPercentage"
+
+    capacity {
+      minimum = "1"
```
Comment (mrlockstar) [inline infrastructure/modules/app-service-plan/autoscale.tf:12]: Maybe parameterise these so its not hard coded

File: infrastructure/modules/app-service-plan/autoscale.tf [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    rule {
+      metric_trigger {
+        metric_name        = "MemoryPercentage"
+        metric_resource_id = azurerm_service_plan.appserviceplan.id
+        time_grain         = "PT1M"
+        statistic          = "Average"
+        time_window        = "PT10M"
+        time_aggregation   = "Average"
+        operator           = "LessThan"
+        threshold          = 25
```
Comment (mrlockstar) [inline infrastructure/modules/app-service-plan/autoscale.tf:46]: Parameterise this.

File: infrastructure/modules/app-service-plan/autoscale.tf [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    rule {
+      metric_trigger {
+        metric_name        = "MemoryPercentage"
+        metric_resource_id = azurerm_service_plan.appserviceplan.id
+        time_grain         = "PT1M"
+        statistic          = "Average"
+        time_window        = "PT10M"
+        time_aggregation   = "Average"
+        operator           = "GreaterThan"
+        threshold          = 70
```
Comment (mrlockstar) [inline infrastructure/modules/app-service-plan/autoscale.tf:26]: Parameterise this.

File: infrastructure/modules/app-service-plan/autoscale.tf (hunk 1 above)
Comment (mrlockstar) [reply infrastructure/modules/app-service-plan/autoscale.tf:12]: And all similar numerical values.

Comment (MacMur85) [reply infrastructure/modules/app-service-plan/autoscale.tf]: Done.

File: infrastructure/modules/app-service-plan/autoscale.tf (hunk 3 above)
Comment (MacMur85) [reply infrastructure/modules/app-service-plan/autoscale.tf]: Done

File: infrastructure/modules/app-service-plan/autoscale.tf (hunk 2 above)
Comment (MacMur85) [reply infrastructure/modules/app-service-plan/autoscale.tf]: Done.

**Comment by MacMur85** [comment]
Improvements implemented, tested and deployed in my test environment.

File: infrastructure/modules/app-service-plan/autoscale.tf (hunk 1 above)
Comment (mrlockstar) [reply infrastructure/modules/app-service-plan/autoscale.tf:12]: I don't see them

**Review by mrlockstar — APPROVED** [review]

**Pull request merged**